homeassistant/components/neato/__init__.py

```diff
@@ -1,197 +1,124 @@
-    "maint_brush_stuck": "Brush stuck",
-    "maint_brush_overload": "Brush overloaded",
-    "maint_bumper_stuck": "Bumper stuck",
-    "maint_customer_support_qa": "Contact customer support",
-    "maint_vacuum_stuck": "Vacuum is stuck",
-    "maint_vacuum_slip": "Vacuum is stuck",
-    "maint_left_drop_stuck": "Vacuum is stuck",
-    "maint_left_wheel_stuck": "Vacuum is stuck",
-    "maint_right_drop_stuck": "Vacuum is stuck",
-    "maint_right_wheel_stuck": "Vacuum is stuck",
-    "not_on_charge_base": "Not on the charge base",
-    "nav_robot_falling": "Clear my path",
-    "nav_no_path": "Clear my path",
-    "nav_path_problem": "Clear my path",
-    "nav_backdrop_frontbump": "Clear my path",
-    "nav_backdrop_leftbump": "Clear my path",
-    "nav_backdrop_wheelextended": "Clear my path",
-    "nav_mag_sensor": "Clear my path",
-    "nav_no_exit": "Clear my path",
-    "nav_no_movement": "Clear my path",
-    "nav_rightdrop_leftbump": "Clear my path",
-    "nav_undocking_failed": "Clear my path",
-}
-
-ALERTS = {
-    "ui_alert_dust_bin_full": "Please empty dust bin",
-    "ui_alert_recovering_location": "Returning to start",
-    "ui_alert_battery_chargebasecommerr": "Battery error",
-    "ui_alert_busy_charging": "Busy charging",
-    "ui_alert_charging_base": "Base charging",
-    "ui_alert_charging_power": "Charging power",
-    "ui_alert_connect_chrg_cable": "Connect charge cable",
-    "ui_alert_info_thank_you": "Thank you",
-    "ui_alert_invalid": "Invalid check app",
-    "ui_alert_old_error": "Old error",
-    "ui_alert_swupdate_fail": "Update failed",
-    "dustbin_full": "Please empty dust bin",
-    "maint_brush_change": "Change the brush",
-    "maint_filter_change": "Change the filter",
-    "clean_completed_to_start": "Cleaning completed",
-    "nav_floorplan_not_created": "No floorplan found",
-    "nav_floorplan_load_fail": "Failed to load floorplan",
-    "nav_floorplan_localization_fail": "Failed to load floorplan",
-    "clean_incomplete_to_start": "Cleaning incomplete",
-    "log_upload_failed": "Logs failed to upload",
-}
-
-
-def setup(hass, config):
+
+async def async_setup(hass, config):
     """Set up the Neato component."""
+
+    if NEATO_DOMAIN not in config:
+        # There is an entry and nothing in configuration.yaml
+        return True
+
+    entries = hass.config_entries.async_entries(NEATO_DOMAIN)
+    hass.data[NEATO_CONFIG] = config[NEATO_DOMAIN]
+
+    if entries:
+        # There is an entry and something in the configuration.yaml
+        entry = entries[0]
+        conf = config[NEATO_DOMAIN]
+        if (
+            entry.data[CONF_USERNAME] == conf[CONF_USERNAME]
+            and entry.data[CONF_PASSWORD] == conf[CONF_PASSWORD]
+            and entry.data[CONF_VENDOR] == conf[CONF_VENDOR]
+        ):
+            # The entry is not outdated
+            return True
+
+        # The entry is outdated
+        hass.config_entries.async_update_entry(entry, data=config[NEATO_DOMAIN])
```

Comment:
Is there a method, that does this? Just checking without creating config entries?

Reply:
No, we need to handle this specifically.

Probably easiest to remove the existing entry and run the flow to create a new one.

@balloob what's the best way to do this?

Reply:
That was my solution before @balloob requested changes

Reply:
I've reverted the procedure to delete outdated entries in order to create new ones. I'd change that again, if we have a better solution

Reply:
We should hear what @balloob suggests. The best thing would be if we can run the config flow but update the existing entry, instead of creating a new, I think. But I haven't seen that yet.

Reply:
@balloob How can we proceed here? I'm sorry to bother you, but I'd love to see these changes within the next release :grinning: according to the calendar on the developers page the beta starts today

Reply:
Can you use the `NeatoConfigFlow.try_login` method to verify the data, and automatically update the config entry if it's ok ?

Reply:
So pretty much use the same functions as the config flow would, but without creating a new config flow.

Reply:
That isn't a solution I would expect. In the config flow we're handling the config and the result is a config entry. That's what we want to do here too.

Could we add an keyword option for a config_entry.entry_id which results in updating an existing entry instead of creating a new entry?

Reply:
Maybe this should be a new entry point to a flow eg `async_step_entry_update`?

And we need a new result type: `RESULT_TYPE_UPDATE_ENTRY`.

Reply:
From what I can see, vey small changes would be needed.

1. We would need to add a case in the finish flow method that checks for the new result type:
https://github.com/home-assistant/home-assistant/blob/f50036772196a1a520f19137be28504ca9edf03f/homeassistant/config_entries.py#L593-L596
2. We need to add the new result type to allowed result types here:
https://github.com/home-assistant/home-assistant/blob/f50036772196a1a520f19137be28504ca9edf03f/homeassistant/data_entry_flow.py#L134-L143
3. We need to add a new callback with the new result type similar as create entry callback:
https://github.com/home-assistant/home-assistant/blob/f50036772196a1a520f19137be28504ca9edf03f/homeassistant/data_entry_flow.py#L203-L222

Reply:
It looks like we're a bit out of the scope of this PR. We have another PR prepared, that refactors some parts of this integration. If we could accept this pr for now, we will adapt the changes in conflg_flow the next time. What do you think?

Reply:
Fine to adopt in a future PR.

The data entry flow framework is also used for options and authentication, so I am a little hesitant to make major changes to it. I still think that this is something that is fine and won't require any changes to the existing Data Entry Flow framework.

```python
if has_config and await validate(config):
    await hass.config_entries.async_update(data=)
```

Reply:
The problem with that is that we might want to redirect to other steps in the config flow, depending on the validation result, just like on the first import of the config.

Reply:
And yes, adding a new result type is definitely out of scope for this PR.

Reply:
> I still think that this is something that is fine and won't require any changes to the existing Data Entry Flow framework.
>
>if has_config and await validate(config):
>    await hass.config_entries.async_update(data=)

Did that. I'll change that behavior in a future PR as soon as we know a preferable solution.